File: 42. Trapping Rain Water/Solution.java
public class Solution {
    public int trap(int[] height) {
        int totalRain = 0;
        int left = 0;
        int right = 0;

        while (right < height.length) {
            if (right - left > 1) {
                while (right < height.length && height[right] < height[left]) {
                    right++;
                }

                if (right >= height.length) {
                    right = height.length - 1;
                    int minH = Math.min(height[left], height[right]);
                    for (int i = right - 1; i > left; i--) {
                        minH = Math.max(minH, height[i]);
                        totalRain += minH - height[i];
                    }
                    break;
                }

                int minH = Math.min(height[left], height[right]);
                System.out.println(left + " " + right + " " + minH);
                for (int i = left; i < right; i++) {
                    totalRain += minH - height[i];
                }
                left = right;
            } else if (height[right] >= height[left]) {
                left = right;
            }
            right++;
        }

        return totalRain;
    }

    public int min(int a, int b) {
        return Math.min(a, b);
    }

    public int max(int a, int b) {
        return Math.max(a, b);
    }


}

File: 42. Trapping Rain Water/Solution.go
package main

func trap(height []int) int {
	totalRain := 0
	left, right := 0, len(height)-1

	for right - left > 1  {
		println(left, right , totalRain)
		if height[left] < height[right] {
			println("left")
			idx := left
			for  idx < right {
				if height[idx] > height[left]  {
					
					break
				}
				totalRain += height[left] - height[idx]
				idx++	
			}
			left = idx
			
		} else {
			println("right")
			i := right
			for  i > left {
				if height[i] > height[right] {
					break
				}
				totalRain += height[right] - height[i]
				i--
			}
			right = i
			
		}
				
	}

    return totalRain
}

func min(a , b int) int {
	if a < b {
		return a
	}
	return b	
}


func main()  {
	testCase := []int{0,1,0,2,1,0,1,3,2,1,2,1}

	println("result 1", trap(testCase))
	
	testCase2 := []int{4,2,3}
	println("result 2", trap(testCase2))
}




